pkg/logql/log/filter.go

```diff
@@ -440,42 +440,71 @@ func contains(line, substr []byte, caseInsensitive bool) bool {
 	return containsLower(line, substr)
 }
 
+// containsLower verifies if substr is a substring of line, with case insensitive comparison.
+// substr is expected to be in lowercase.
 func containsLower(line, substr []byte) bool {
 	if len(substr) == 0 {
 		return true
 	}
 	if len(substr) > len(line) {
 		return false
 	}
-	j := 0
-	for len(line) > 0 {
-		// ascii fast case
-		if c := line[0]; c < utf8.RuneSelf && substr[j] < utf8.RuneSelf {
-			if c == substr[j] || c+'a'-'A' == substr[j] || c == substr[j]+'a'-'A' {
-				j++
-				if j == len(substr) {
-					return true
```

Comment:
Interesting I've always thought Loki was using `strings.Index` but I guess one cannot ignore character casing there.

Anyhow, during my work on a SIMD [substring search](https://github.com/jeschkies/go-memmem/) I've found that the underlying Rabin-Karp algorithm is soo much faster. I'm curious if it could be adapted to be case-insensitive 🤷

Reply:
We don't for case sensitive yes. What do you think of the PR though that would help me to get a review.